consensus/notify/src/notification.rs

```diff
@@ -0,0 +1,173 @@
+use consensus_core::{block::Block, tx::TransactionId, utxo::utxo_diff::UtxoDiff};
+use derive_more::Display;
+use hashes::Hash;
+use kaspa_notify::{
+    events::EventType,
+    full_featured,
+    notification::Notification as NotificationTrait,
+    subscription::{
+        single::{OverallSubscription, UtxosChangedSubscription, VirtualChainChangedSubscription},
+        Single,
+    },
+};
+use std::sync::Arc;
+
+full_featured! {
+#[derive(Clone, Debug, Display)]
+pub enum Notification {
+    #[display(fmt = "BlockAdded notification: block hash {}", "_0.block.header.hash")]
+    BlockAdded(BlockAddedNotification),
+
+    #[display(fmt = "VirtualChainChanged notification: {} removed blocks, {} added blocks, {} accepted transactions", "_0.removed_chain_block_hashes.len()", "_0.added_chain_block_hashes.len()", "_0.accepted_transaction_ids.len()")]
+    VirtualChainChanged(VirtualChainChangedNotification),
+
+    #[display(fmt = "FinalityConflict notification: violating block hash {}", "_0.violating_block_hash")]
+    FinalityConflict(FinalityConflictNotification),
+
+    #[display(fmt = "FinalityConflict notification: violating block hash {}", "_0.finality_block_hash")]
+    FinalityConflictResolved(FinalityConflictResolvedNotification),
+
+    #[display(fmt = "UtxosChanged notification")]
+    UtxosChanged(UtxosChangedNotification),
+
+    #[display(fmt = "SinkBlueScoreChanged notification: virtual selected parent blue score {}", "_0.sink_blue_score")]
+    SinkBlueScoreChanged(SinkBlueScoreChangedNotification),
+
+    #[display(fmt = "VirtualDaaScoreChanged notification: virtual DAA score {}", "_0.virtual_daa_score")]
+    VirtualDaaScoreChanged(VirtualDaaScoreChangedNotification),
+
+    #[display(fmt = "PruningPointUtxoSetOverride notification")]
+    PruningPointUtxoSetOverride(PruningPointUtxoSetOverrideNotification),
+
+    #[display(fmt = "NewBlockTemplate notification")]
+    NewBlockTemplate(NewBlockTemplateNotification),
+}
+}
+
+impl NotificationTrait for Notification {
+    fn apply_overall_subscription(&self, subscription: &OverallSubscription) -> Option<Self> {
+        match subscription.active() {
+            true => Some(self.clone()),
+            false => None,
+        }
+    }
+
+    fn apply_virtual_chain_changed_subscription(&self, subscription: &VirtualChainChangedSubscription) -> Option<Self> {
+        match subscription.active() {
+            true => {
+                if let Notification::VirtualChainChanged(ref payload) = self {
+                    if !subscription.include_accepted_transaction_ids() && !payload.accepted_transaction_ids.is_empty() {
```

Comment:
Maybe explain in a comment that if no subscription + `payload.accepted_ids` is not empty, then we must re-create the object and drop the ids, otherwise we can clone it as is